### DISCUSSION TITLE
Infering data type from context.db or context.query

### DISCUSSION THREAD

**Discussion by adrianbienias** [keystonejs/keystone#8484] [Questions] [answered]
By using context API to query data, I expect it to behave similarly to Prisma API that provides queried data types.

Examples:

`const item = await context.db.<schema_model>.findOne({ where: { id: <item_id> })`
gives
`BaseItem | null`

`const item = await context.query.<schema_model>.findOne({ where: { id: <item_id> }, query: "<some_field>" })`
gives
`Record<string, any>`

In both cases I can call
`item.<non_existing_field>`
without being guarded by TypeScript

How to get proper types?

**Comment by pahaz** [comment]
I have the same question. There are no doc pages related to types and the Keystone6 type system.

**Comment by mmachatschek** [comment, marked as answer]
The list needs to be typed with the correct context.

Here is a link to one of the examples, with this the `context.(db|prisma)` api should have the corresponding types loaded:

https://github.com/keystonejs/keystone/blob/7e0d592396e45f789a18bc25edce36e0a8d2c8c1/examples/usecase-blog/schema.ts#L12-L16

Without seeing more of your code, I suppose you export a list in a separate file and import it to the `keystone.ts` config.

If you're using the `getContext` or any other context that seems to be missing the types, you may set the type to `Context` like so:

https://github.com/keystonejs/keystone/blob/7e0d592396e45f789a18bc25edce36e0a8d2c8c1/examples/nextjs-keystone/src/keystone/context.ts#L3-L8

or like this https://github.com/keystonejs/keystone/discussions/8484#discussioncomment-5624676

**Comment by adrianbienias** [comment]
I was using context inside [extendExpressApp](https://keystonejs.com/docs/config/config#extend-express-app) as
```
const context = await commonContext.withRequest(req, res)
```
and that led to having `KeystoneContext<BaseKeystoneTypeInfo>` type for context

Casting it as `Context` solved the problem
```
const context = (await commonContext.withRequest(req, res)) as Context
```

**Reply by dcousens** [reply]
@adrianbienias `commonContext` should be of type `Context`, [you might need to set `config<TypeInfo>`](https://github.com/keystonejs/keystone/blob/d33bb4764669db22683b30edb8dcf8e3c4e9b3fc/examples/usecase-versioning/keystone.ts#L4-L6).

If you wanted to update the example in https://github.com/keystonejs/keystone/blob/main/examples/extend-express-app/keystone.ts that could be helpful for others too! :yellow_heart:

**Reply by adrianbienias** [reply]
I can create a PR for that example, but what about other examples?
Shouldn't `config<TypeInfo>` be updated in all examples?

**Reply by dcousens** [reply]
@adrianbienias `TypeInfo`, while complete, isn't necessarily helpful for your types if you aren't writing the inferred code inline.
Probably we should write `extendExpressApp` as a free function with a `: Context` parameter...

**Reply by tilman** [reply]
> I can create a PR for that example, but what about other examples? Shouldn't `config<TypeInfo>` be updated in all examples?

Have you already opened the PR? Seems like I ran into the exact same issue and was hoping to save me some work :)

**Reply by adrianbienias** [reply]
I haven't. Manually adding `<TypeInfo>` is not much work, I guess ;)

**Reply by adrianbienias** [reply]
https://github.com/keystonejs/keystone/pull/8500

**Comment by dcousens** [comment]
@mmachatschek's answer covered using `lists: Lists` as a type refinement pattern to refine the types of `context.db` and `context.prisma`. Unfortunately, `context.query` is still left with `any` types.

I have opened the following discussion for `context.query`, https://github.com/keystonejs/keystone/discussions/8498
